File: 2022/day08/part2.py
#!/usr/bin/env python3


def is_visible(x: int, y:int, map: list[str]) -> int:
    right: int = 0
    for mouving_x in range (x + 1, len(map[y])):
        right += 1
        if map[y][x] <= map[y][mouving_x]:
            break
    left: int = 0
    for mouving_x in range (x - 1, -1, -1):
        left += 1
        if map[y][x] <= map[y][mouving_x]:
            break
    bottom: int = 0
    for mouving_y in range (y + 1, len(map)):
        bottom += 1
        if map[y][x] <= map[mouving_y][x]:
            break
    top: int = 0
    for mouving_y in range (y - 1, -1, -1):
        top += 1
        if map[y][x] <= map[mouving_y][x]:
            break
    return right * left * bottom * top


file = open("./input.txt", "r")
my_file_parse: list[str] = file.readlines()
file.close()

my_file_parse = [line[:-1] for line in my_file_parse]

for y in range (0, len(my_file_parse)):
    print(my_file_parse[y])

total: int = 0

for y in range (1, len(my_file_parse) - 1):
    for x in range (1, len(my_file_parse) - 1):
        tmp_total = is_visible(x, y, my_file_parse)
        if tmp_total > total:
            total = tmp_total
print(total)

# atempt 1 517440 is the result

File: 2024/day04/part2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

// Define offsets for X-MAS pattern
var xmasOffsets = [4][2]int{
	{-1, -1}, // Top-left MAS
	{1, -1},  // Bottom-left MAS
	{-1, 1},  // Top-right MAS
	{1, 1},   // Bottom-right MAS
}

func countXMAS(grid []string) int {
	height := len(grid)
	if height == 0 {
		return 0
	}

	width := len(grid[0])
	count := 0

	for y := 1; y < height-1; y++ {
		for x := 1; x < width-1; x++ {
			// Check if the center is 'A'
			if grid[y][x] != 'A' {
				continue
			}

			// Check each diagonal pair
			foundXMAS := 0
			for _, offset := range xmasOffsets {
				dx, dy := offset[0], offset[1]
				nx, ny := x+dx, y+dy
				ox, oy := x-dx, y-dy

				// Ensure bounds and validate 'M' and 'S' pairing
				if ny >= 0 && ny < height && nx >= 0 && nx < width &&
					oy >= 0 && oy < height && ox >= 0 && ox < width &&
					((grid[ny][nx] == 'M' && grid[oy][ox] == 'S') || (grid[ny][nx] == 'S' && grid[oy][ox] == 'M')) {
					foundXMAS++
				}
			}

			// A valid X-MAS requires exactly 4 diagonal pairs
			if foundXMAS == 4 {
				count++
			}
		}
	}

	return count
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	var grid []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) > 0 {
			grid = append(grid, line)
		}
	}

	if err := scanner.Err(); err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	if len(grid) == 0 {
		fmt.Println("Grid is empty.")
		return
	}

	result := countXMAS(grid)
	fmt.Println("Result:", result)
}
